File: c/Find Kth Bit in Nth Binary String.c
#include <stdio.h>

char findKthBitHelper(int n, int k) {
    if (n == 1) {
        return '0';  
    }
    int length = (1 << n) - 1;  
    int mid = (length + 1) / 2; 

    if (k == mid) {
        return '1';
    } else if (k < mid) {
        return findKthBitHelper(n - 1, k);  
    } else {
        char bit = findKthBitHelper(n - 1, length - k + 1);
        return bit == '0' ? '1' : '0';  
    }
}

char findKthBit(int n, int k) {
    return findKthBitHelper(n, k);
}

// Test case
int main() {
    // case 1
    int n1 = 3, k1 = 1;
    printf("Test case 1 - Expected: 0, Got: %c\n", findKthBit(n1, k1));
    
    // case 2
    int n2 = 4, k2 = 11;
    printf("Test case 2 - Expected: 1, Got: %c\n", findKthBit(n2, k2));
    
    // Additional cases
    int n3 = 2, k3 = 3;
    printf("Test case 3 - Expected: 1, Got: %c\n", findKthBit(n3, k3));
      
    return 0;
}

File: JavaCode/FindKthBitInNthBinaryString.java
public class FindKthBitInNthBinaryString {
    public static void main(String[] args) {
        Solution sol = new Solution();

        // case 1
        int n1 = 3;
        int k1 = 1;
        System.out.println("Result of test case 1 : " + sol.findKthBit(n1, k1)); // Expected "0"

        // case 2
        int n2 = 4;
        int k2 = 11;
        System.out.println("Result of test case 2 : " + sol.findKthBit(n2, k2)); // Expected "1"
    }
}

class Solution {

    public char findKthBit(int n, int k) {
        return find(n, k);
    }

    private char find(int n, int k) {
        if (n == 1) {
            return '0';
        }

        int mid = 1 << (n - 1);   // midpoint of current sequence

        if (k == mid) {
            return '1';
        }

        if (k < mid) {
            return find(n - 1, k);
        }

        // mirrored position
        int mirrored = (1 << n) - k;

        // invert the mirrored result
        return (char) (find(n - 1, mirrored) ^ 1);
    }
}
